### ISSUE TITLE
Support for embedding test code within the application

### ISSUE THREAD

**Issue by mildred** [philsquared/Catch#187]
It would be nice if it was possible to embed the test code in the same files that are tested. For example, in the class .cpp file, I could have implementation for the methods of the class and at the same case the test cases of the class.

It is actually possible to build a test executable doing this, but it is not possible to build an application that way. Some symbols defined along main are not found at link time.

I would imagine:
- a debug build of the object files of the application that contain the test code
- a debug build of the main() function of the application that provides dummy implementation for the symbols necessary for Catch, but otherwise start the application as usual
- a debug build of the test executable that contains the main() that will execute the tests
- a release build of the object files of the application, stripped of the test code
- a release build of the main() function of the application that is not required to provide the dummy implementation anymore because no test cases are compiled in.

basically, it would require:
- defining CATCH_CONFIG_NOOP turn the test cases into inline static functions that would be optimized away, and not referenced by Catch. This is for the release build.
- defining CATCH_CONFIG_DUMMY_MAIN will instanciate all objects necessary for the tests with an empty implementation to allow running the application when tests are compiled in and we don't want to run them.
